<a href="https://colab.research.google.com/github/Vishnukvsvk/ML-Foundation-Google-Developers-Youtube/blob/master/TF_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this exercise I tried to build a neural network that predicts the price of a house according to a simple formula.

So, imagine if house pricing was as easy as a house costs 50k + 50k per bedroom, so that a 1 bedroom house costs 100k, a 2 bedroom house costs 150k etc.

How would you create a neural network that learns this relationship so that it would predict a 7 bedroom house as costing close to 400k etc.

Hint: Your network might work better if you scale the house price down. You don't have to give the answer 400...it might be better to create something that predicts the number 4, and then your answer is in the 'hundreds of thousands' etc.

In [0]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

The input shape<br>
In Keras, the input layer itself is not a layer, but a tensor. It's the starting tensor you send to the first hidden layer. This tensor must have the same shape as your training data. Example: if you have 30 images of 50x50 pixels in RGB (3 channels), the shape of your input data is (30,50,50,3)

In [0]:
model = tf.keras.Sequential([keras.layers.Dense(units=1,input_shape=[1])]) #units = no. of nuerons

**Compile**

In [0]:
model.compile(optimizer='sgd',loss='mean_squared_error')

y=50k+x*50k  ==> y = (1+x)*50k

In [0]:
xs = np.array([1.0,2.0,3.0,4.0,5.0],dtype=float)
ys = np.array([2.0,3.0,4.0,5.0,6.0],dtype=float)

In [8]:
model.fit(xs,ys,epochs=100)

Epoch 1/100
1/1 [==============================] - 0s 1ms/step - loss: 2.8934
Epoch 2/100
1/1 [==============================] - 0s 1ms/step - loss: 1.7428
Epoch 3/100
1/1 [==============================] - 0s 2ms/step - loss: 1.0720
Epoch 4/100
1/1 [==============================] - 0s 1ms/step - loss: 0.6806
Epoch 5/100
1/1 [==============================] - 0s 1ms/step - loss: 0.4522
Epoch 6/100
1/1 [==============================] - 0s 2ms/step - loss: 0.3187
Epoch 7/100
1/1 [==============================] - 0s 2ms/step - loss: 0.2406
Epoch 8/100
1/1 [==============================] - 0s 6ms/step - loss: 0.1946
Epoch 9/100
1/1 [==============================] - 0s 3ms/step - loss: 0.1675
Epoch 10/100
1/1 [==============================] - 0s 1ms/step - loss: 0.1513
Epoch 11/100
1/1 [==============================] - 0s 3ms/step - loss: 0.1415
Epoch 12/100
1/1 [==============================] - 0s 849us/step - loss: 0.1355
Epoch 13/100
1/1 [==============================] - 0s 2ms/

In [10]:
print(model.predict([7.0]))

[[8.582962]]


In [11]:
model.fit(xs,ys,epochs=500)

Epoch 1/500
1/1 [==============================] - 0s 1ms/step - loss: 0.0702
Epoch 2/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0697
Epoch 3/500
1/1 [==============================] - 0s 4ms/step - loss: 0.0692
Epoch 4/500
1/1 [==============================] - 0s 1ms/step - loss: 0.0688
Epoch 5/500
1/1 [==============================] - 0s 3ms/step - loss: 0.0683
Epoch 6/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0678
Epoch 7/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0674
Epoch 8/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0669
Epoch 9/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0665
Epoch 10/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0660
Epoch 11/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0656
Epoch 12/500
1/1 [==============================] - 0s 2ms/step - loss: 0.0651
Epoch 13/500
1/1 [==============================] - 0s 2ms/st

In [13]:
print(model.predict([7]))

[[8.107209]]


Actual answer is **8**. It predicted **8.5** for **100** epochs and **8.1** for **500** epochs.<br>

As epochs increases,loss decreases.